### ISSUE TITLE
APIs 2.0

### ISSUE THREAD

**Issue by kristjanmik** [kristjanmik/apis#80]
We have to rethink the project...

I just had a great meeting with @arnorhs and we were discussing the future of the platform. We came up with some interesting thoughts and we truly believe we can grow this project beyond what it is today into something bigger than a simple JSON scraper.

We have to start to decouple the project more than we are currently doing. All the endpoints are hosted in the endpoints directory and there is a circular dependency tree at this point. A cool thing would be to create one npm package for each endpoint and then a pull request into the project would not be needed to update the endpoint itself. The module would need to export its tests and documentation.
By doing this we are decoupling the project so we can start to focus more on the core development and give the power back to the endpoint creators(and the ownership).

When we have split the project into more maintainable parts we can really start to focus on the core itself. Me and @arnorhs discussed some features that would be nice to have in the platform but we definitely have to prioritise them in terms of implementation stages. 

Things that could benefit the platform:

Caching layer
Each endpoint can cache its own output or store some values in the caching layer. We currently offer output caching but it does not support caching headers at this point.

Database layer
Each endpoint would be able to store some data in a document storage such as MongoDB. The endpoint could then read data from the database and do something more than just realtime stuff. A cool idea would be to create a hashing helper that could hash a scraped content from a website and look up if the has exists in the database.

Worker and Queue layer
Each endpoint would be able to create a task that would execute on a specific interval or on a event basis. The endpoint could then start to aggregate data outside the requests that are currently happening and could use the database/caching layer to dump the data into.

Pubsub
With a pubsub model we are able to accept connections that go beyond the current connection model. We would be able to offer hooks on specific events or if the connection is long living we could pump out notifications through that (e.g. twitter firehose). For this to be able to work we need a working worker layer and some persistence/queue layer.

Notification layer
When something goes down or a state alters on the main server we could send out emails to the people that subscribe to them

What are your thoughts on this? This is just the initial brain dump after the meeting and I think that we need to have a good discussion before we can make some drastic changes.

Cheers!
@kristjanmik
